omegaconf/base.py

```diff
@@ -404,13 +404,20 @@ def _resolve_interpolation_from_parse_tree(
     ) -> Optional["Node"]:
         from .nodes import StringNode
 
-        resolved = self.resolve_parse_tree(
-            parse_tree=parse_tree,
-            key=key,
-            parent=parent,
-            throw_on_missing=throw_on_missing,
-            throw_on_resolution_failure=throw_on_resolution_failure,
-        )
+        try:
```

Comment:
I believe it's impossible to process a string interpolation that is interpolating into missing of unresolved interpolations.
What is the purpose of returning None in such cases?
I am not sure the two flags
```
        throw_on_missing: bool,
        throw_on_resolution_failure: bool,
```

Even makes sense here. shouldn't they always be true?

Reply:
So for instance `ValueNode._is_none()` looks like this, and would break without these flags set to False:
```python
    def _is_none(self) -> bool:
        if self._is_interpolation():
            node = self._dereference_node(
                throw_on_resolution_failure=False, throw_on_missing=False
            )
            if node is None:
                # missing or resolution failure
                return False
        else:
            node = self
        return node._value() is None
```
That being said, this made me look into other parts of the code where I believe some adjustments are needed. I'll follow up on this.

Reply:
> That being said, this made me look into other parts of the code where I believe some adjustments are needed. I'll follow up on this.

So:
* c13fe71 simplifies `_is_missing()` (and fixes it for containers so that interpolated containers pointing to a missing variable aren't missing anymore)
* 74a92d8 removes a bunch of `throw_on_missing` arguments that didn't make sense anymore (related to your point -- I think). The idea is that the `throw_on_missing` argument passed to `_dereference_node()` should only be used if the node isn't an interpolation, to decide whether or not to throw an exception if the node is missing. But if it's an interpolation, it's not missing, and thus it shouldn't be used (instead, if a missing value is encountered while resolving the interpolation, then it should be considered a resolution failiure)

I still kept `throw_on_resolution_failure` since it's needed in multiple places (e.g., during a merge) when attempting to dereference a node without raising an exception in case of failure.

Reply:
Also fixed coverage in 9ec934d (turns out that there were some situations that became now impossible to reach, due to interpolations not being missing anymore)

Reply:
> We can change the semantics of is_none() to follow the same logic:
> is_none is true if and only if the value is None. Interpolation is not None.

We should probably separate the internal `_is_none()` from the public `OmegaConf.is_none()`.
* For the internal `_is_none()`, I would need to double check again to be 100% sure, but I believe it can indeed be changed to not resolve interpolations
* For the public `OmegaConf.is_none()`, maybe it would be best to deprecate it to avoid confusion, telling users they should use `cfg.x is None` instead. The only potential issue is that it doesn't cover the case where someone would call `OmegaConf.is_none(node)` (without a key). If we want to keep supporting that, maybe we could add a new `OmegaConf.resolve(node)` that would return the (resolved) value of the node? (so this would become `OmegaConf.resolve(node) is None`)

Reply:
Ok sounds good, I created #547 for this (I'd rather do it in a separate PR)